### DISCUSSION TITLE
Serializing/deserializing structures with pointers (swizzling)

### DISCUSSION THREAD

**Discussion by mqnc** [nlohmann/json#3406] [Q&A]
I essentially want the same thing as this person: https://github.com/nlohmann/json/discussions/2916

Here is my approach:
- Before serializing, I create an empty pointee json object, serving as an address map.
- Whenever I encounter a pointer in the object structure that I'm serializing, I create a new entry in the address map, using the pointed memory address as key, except if an entry with that key is already there. Then I store a json pointer to that map entry.

Example:
```json
{
    "__map__": {
        "0x1234": {
            "name": "Wesley",
            "parent": "/__map__/0x1337"
        },
        "0x1337": {
            "name": "Beverly",
            "child": "/__map__/0x1234"
        }
    },
    "family": ["/__map__/0x1234", "/__map__/0x1337"]
}
```
This would be the result of serializing the `family` array, which could be a vector of two pointers.

My only issue is, every `to_json` function must know the address map. It would work to store it in a global variable (which could be locked by a mutex or something) but that is not very elegant because global variables are frowned upon by the naysayers. On the other hand, handing around a reference as a third optional parameter for the `to_json` function probably entails a lot of work on the library side and would probably not work with the macros and the implicit casts. Or I can write a bunch of `to_json_graph(&j, &o, &map)` functions for all objects containing pointers somewhere down the line and I have to call this one explicitly... (tbh I just came up with this third option while writing this and I think it looks the cleanest and I'm probably going to stick with it).

Are there any better options?

**Comment by gregmarr** [comment]
Another possible solution is to have every object in your tree know about or at least be able to get to the address map, such as there being a single top level object at the top of the graph, and every object has a pointer to their parent which would eventually lead to this top level object.  That would at least allow you to implement this without changes to the library, and would be cleaner than the global object.  This could require significant changes to your object model, however.

**Reply by sancelot** [reply]
not really clear without a concrete example
